devs/docs/logging.rst

```diff
@@ -0,0 +1,45 @@
+=======
+Logging
+=======
+
+Debug/Trace Logging
+...................
+
+When working with asynchronous programming very often it's quite hard to follow
+execution flow or identify a possible bug. A common code area that things can
+get quite tricky is the distributed execution engine.
+
+Therefore, it is advisable, when working on such components (implementing new
+features, refactoring code or maybe just debugging) to try to introduce helpful
+log messages. Those messages will not only help with the current work but will
+be extremely useful for future work on those components.
+
+Such messages should be logged in ``DEBUG`` or ``TRACE`` level depending on how
+fine grain information they provide. To avoid introducing performance overhead
+they must also be "guarded" by checking the enabled level of the logger, e.g.::
+
+  if (logger.isTraceEnabled()) {
```

Comment:
Might want to note that this should ideally be done only once if the component is initialized, unless it's a singleton.

`isXEnabled` itself is relative expensive.

Also, maybe link to https://www.brandur.org/logfmt as a guideline for how to format the log statements.

Reply:
Need to rephrase it yes, the statement should be guarded if does some "computation" or has more arguments than the defined methods and we fallback to the vararg method signature.